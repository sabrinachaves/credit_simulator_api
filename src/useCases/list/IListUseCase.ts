import { ICreditSimulation } from '@domain/entities/CreditSimulation';

export interface IListCreditsFilters {
  minAmount?: number;
  maxAmount?: number;
  createdAfter?: Date;
  createdBefore?: Date;
}

export default interface IListUseCase {
  handle(filters?: IListCreditsFilters, page?: number, pageSize?: number): Promise<ICreditSimulation[]>;
}
