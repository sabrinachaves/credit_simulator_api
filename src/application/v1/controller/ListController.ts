import { buildErrorInfo } from '@infrastructure/parser/ErrorInfo';
import IListUseCase, { IListCreditsFilters } from '@useCases/list/IListUseCase';
import { OK } from 'http-status';
import { Request, Response } from 'express';

export default class ListController {
  constructor(private listUseCase: IListUseCase) {}

  public execute = async (request: Request, response: Response) => {
    try {
      const filters: IListCreditsFilters = request.query;
      const page = request.query.page ? Number(request.query.page) : undefined;
      const pageSize = request.query.pageSize ? Number(request.query.pageSize) : undefined;

      const creditSimulations = await this.listUseCase.handle(filters, page, pageSize);

      return response.status(OK).json(creditSimulations);
    } catch (err: any) {
      const errorInfo = buildErrorInfo(err);

      console.error('failed to create a credit simulation', errorInfo);
      return response.status(errorInfo.code).json({ error: errorInfo.errorMessage });
    }
  };
}
