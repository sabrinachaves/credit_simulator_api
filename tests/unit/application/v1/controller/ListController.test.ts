import { Request } from 'express';
import ListController from '../../../../../src/application/v1/controller/ListController';
import { INTERNAL_SERVER_ERROR, OK } from 'http-status';
import InternalServerError from '../../../../../src/domain/exceptions/InternalServerError';
import ListUseCaseMock from '../../../../mocks/ListUseCaseMock';
import { CreditSimulationMock } from '../../../../mocks/CreditMock';
import IListUseCase from '../../../../../src/useCases/list/IListUseCase';

describe('ListController', () => {
  const listUseCase = new ListUseCaseMock();

  it('should handle the request #unit', async () => {
    const instance = new ListController(listUseCase);
    const req = { query: {} } as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    listUseCase.handle = jest.fn().mockResolvedValue([{ ...CreditSimulationMock }]);

    await instance.execute(req, res);

    expect(listUseCase.handle).toHaveBeenCalledWith({}, undefined, undefined);
    expect(res.status).toHaveBeenCalledWith(OK);
  });

  it('should handle the request with query params #unit', async () => {
    const instance = new ListController(listUseCase);
    const req = {
      query: {
        minAmount: 5000,
        page: 2,
        pageSize: 15,
      },
    } as unknown as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    listUseCase.handle = jest.fn().mockResolvedValue([{ ...CreditSimulationMock }]);

    await instance.execute(req, res);

    expect(listUseCase.handle).toHaveBeenCalledWith(expect.objectContaining({ minAmount: 5000 }), 2, 15);
    expect(res.status).toHaveBeenCalledWith(OK);
  });

  it('should throw an error when useCase throws #unit', async () => {
    const instance = new ListController(listUseCase);
    const req = { query: {} } as Request;
    const res: any = {
      status: jest.fn().mockImplementation(() => ({
        json: jest.fn(),
      })),
    };

    listUseCase.handle = jest.fn().mockRejectedValueOnce(new InternalServerError('Internal Server Error'));

    await instance.execute(req, res);

    expect(listUseCase.handle).toHaveBeenCalled();
    expect(res.status).toHaveBeenCalledWith(INTERNAL_SERVER_ERROR);
  });

  it('should match the snapshot #unit', () => {
    const instance = new ListController({} as IListUseCase);
    expect(instance).toMatchSnapshot();
  });
});
