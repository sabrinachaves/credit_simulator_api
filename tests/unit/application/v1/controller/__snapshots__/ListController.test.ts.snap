// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListController should match the snapshot #unit 1`] = `
ListController {
  "execute": [Function],
  "listUseCase": Object {},
}
`;
